src/ghostwriter/models/v1/mapping.py

```diff
@@ -0,0 +1,180 @@
+"""Definition of the mapping from a top-level route to its substituted form,
+as well as any hooks that should be called before passing the input route
+to be rewritten.
+"""
+
+from __future__ import annotations
+
+from collections.abc import Awaitable, Callable
+from string import Template
+from typing import Annotated
+
+from pydantic import AfterValidator, BaseModel, BeforeValidator, Field
+
+from ... import hooks
+from ...exceptions import (
+    HookError,
+    HookNotFoundError,
+    MatchNotFoundError,
+    ResolutionError,
+)
+from ..substitution import Parameters
+
+
+def canonicalize_source_route(v: str) -> str:
+    """Force source route to have one leading and one trailing slash."""
+    return f"/{v.strip('/')}/"
+
+
+def resolve_hooks(
+    v: list[str] | list[Callable[[Parameters], Awaitable[None]]],
+) -> list[Callable[[Parameters], Awaitable[None]]]:
+    """Hooks will be listed as strings in the config file.  This loads
```

Comment:
Not sure if you plan on adding docstrings, but might help out with being able to quickly identify what the params should be and how this works. 
For this, one maybe also changing `v` to something more descriptive, and I'm also thinking that this is not just resolving but also loading, so perhaps naming it `load_hooks`.

Suggestion snippet:
    
    def load_hooks(
            v: list[str] | list[Callable[[Parameters], Awaitable[None]]],
    ) -> list[Callable[[Parameters], Awaitable[None]]]:
        """Hooks will be listed as strings in the config file.  This loads
        the corresponding functions.
    
        Parameters
        ----------
        v : list[str] | list[Callable[[Parameters], Awaitable[None]]]
            List of hooks to be loaded.  If the hook is a string, it is
            assumed to be a function in the ghostwriter.hooks module.
    
        Returns
        -------
        list[Callable[[Parameters], Awaitable[None]]]
            List of hooks to be called in order before substituting and returning
            the route.
    
        Raises
        ------
        HookNotFoundError
            If the hook could not be loaded.
        """

Reply:
Well, the `v` is because that's canonical for Pydantic model validators.  I've added some commentary (and renamed it to load_hooks) to indicate that it is a model validator, whose intention is to read the list of hook names supplied in the config and turn those into the corresponding Python objects.